blockstack-vm/src/types.rs

```diff
@@ -0,0 +1,213 @@
+                       data_map: data_map })
+
+    }
+
+    pub fn get(&self, name: &str) -> InterpreterResult {
+        if let Some(value) = self.data_map.get(name) {
+            Ok(value.clone())
+        } else {
+            Err(Error::InvalidArguments(format!("No such field {:?} in tuple", name)))
+        }
+        
+    }
+}
+
+impl TypeSignature {
+    pub fn new(atomic_type: AtomTypeIdentifier, dimension: u8) -> TypeSignature {
+        TypeSignature { atomic_type: atomic_type,
+                        dimension: dimension }
+    }
+
+    pub fn type_of(x: &Value) -> TypeSignature {
+        match x {
+            Value::Void => TypeSignature::new(AtomTypeIdentifier::VoidType, 0),
+            Value::Int(_v) => TypeSignature::new(AtomTypeIdentifier::IntType, 0),
+            Value::Bool(_v) => TypeSignature::new(AtomTypeIdentifier::BoolType, 0),
+            Value::Buffer(_v) => TypeSignature::new(AtomTypeIdentifier::BufferType, 0),
+            Value::List(_v, type_signature) => type_signature.clone(),
+            Value::Tuple(v) => TypeSignature::new(AtomTypeIdentifier::TupleType(
+                v.type_signature.clone()), 0)
+        }
+    }
+
+    pub fn get_list_type_for(x: &Value) -> Result<TypeSignature, Error> {
+        match x {
+            Value::Void => Err(Error::InvalidArguments("Cannot construct list of void types".to_string())),
+            Value::Tuple(_a) => Err(Error::InvalidArguments("Cannot construct list of tuple types".to_string())),
+            _ => {
+                let mut base_type = TypeSignature::type_of(x);
+                base_type.dimension += 1;
+                Ok(base_type)
+            }
+        }
+    }
+
+    fn get_atom_type(typename: &str) -> Result<AtomTypeIdentifier, Error> {
+        match typename {
+            "int" => Ok(AtomTypeIdentifier::IntType),
+            "void" => Ok(AtomTypeIdentifier::VoidType),
+            "bool" => Ok(AtomTypeIdentifier::BoolType),
+            "buff" => Ok(AtomTypeIdentifier::BufferType),
+            _ => Err(Error::ParseError(format!("Unknown type name: '{:?}'", typename)))
+        }
+    }
+
+    
+    fn get_list_type(prefix: &str, typename: &str, dimension: &str) -> Result<TypeSignature, Error> {
+        if prefix != "list" {
+            let message = format!("Unknown type name: '{}-{}-{}'", prefix, typename, dimension);
+            return Err(Error::ParseError(message))
+        }
+        let atom_type = TypeSignature::get_atom_type(typename)?;
+        let dimension = match u8::from_str_radix(dimension, 10) {
+            Ok(parsed) => Ok(parsed),
+            Err(_e) => Err(Error::ParseError(
+                format!("Failed to parse dimension of type: '{}-{}-{}'",
+                        prefix, typename, dimension)))
+        }?;
+        Ok(TypeSignature::new(atom_type, dimension))
+    }
+
+    pub fn parse_type_str(x: &str) -> Result<TypeSignature, Error> {
```

Comment:
How long can this string get?  Asking because if we're going to `split` it, we might end up doing `O(poly(n))` work for `n` occurrences of `-`.

Reply:
Won't split only ever do `O(n)` work ?

Reply:
Depending on the allocator implementation, it could do `O(n)` `malloc()`s, which each could take `O(log m)` time (for `m` blocks allocated).  Not sure what the actual implementation does, but I'm of the school of thought that we should prepare for the worst.

Reply:
Okay, yes, avoiding non-constant object instantions is a good idea. Will just switch to using `splitn` here.